### DISCUSSION TITLE
integrations/neural-magic/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8506] [Docs]
# integrations/neural-magic/

Learn how to deploy your YOLOv8 models rapidly using Neural Magic’s DeepSparse. This guide focuses on integrating Ultralytics YOLOv8 with the DeepSparse Engine for high-speed, CPU-based inference, leveraging advanced neural network sparsity techniques.

https://docs.ultralytics.com/integrations/neural-magic/

**Comment by sanvir19** [comment]
"Can I use DeepSparse on my Jetson Nano Dev Kit? Do I need to convert my YOLO model to TensorRT for that? The blog mentions converting it to ONNX format. Will using DeepSparse make my model run faster on my Jetson?"

**Reply by glenn-jocher** [reply]
@sanvir19 hey there! 😊

DeepSparse is designed to optimize neural network execution on CPUs, leveraging techniques like sparsity to enhance performance. While it's great for CPU environments, the Jetson Nano Dev Kit primarily benefits from GPU acceleration.

For your Jetson Nano, converting your YOLO model to TensorRT is indeed the recommended path. TensorRT is specifically designed to optimize deep learning inference on NVIDIA GPUs, including those on Jetson devices. Exporting your model to ONNX is a step in the right direction, as ONNX models can be further converted to TensorRT for deployment on Jetson.

Using DeepSparse could potentially make your model run faster on CPU-based systems, but for Jetson Nano, focusing on TensorRT will leverage the GPU for optimal performance. Here's a quick guide to convert your model to TensorRT:

1. Export your YOLO model to ONNX format:
   ```bash
   yolo export model=yolov8n.pt format=onnx
   ```

2. Use the `onnx2trt` tool to convert the ONNX model to TensorRT:
   ```bash
   onnx2trt yolov8n.onnx -o yolov8n.trt
   ```

This approach should give you the best inference speed on your Jetson Nano. If you have any more questions or need further assistance, feel free to ask!

**Comment by Hemilibeatriz** [comment]
Which version of YOLO was DeepSparse tested on? In version 8.2.2, I'm unable to get it to work

**Reply by glenn-jocher** [reply]
Hi there! It seems you're having trouble with DeepSparse on YOLOv8 version 8.2.2. The integration and testing with Neural Magic’s DeepSparse were primarily conducted on earlier versions of YOLOv8. It's possible that there have been changes in version 8.2.2 that affect compatibility.

I recommend checking the version compatibility in the Neural Magic documentation or trying to export and run your model with an earlier version of YOLOv8. If the issue persists, it might be helpful to reach out to Neural Magic's support or check their community forums for more specific guidance.

Here’s a quick example of how you might export your model to ONNX, which is required for DeepSparse:

```bash
yolo task=detect mode=export model=yolov8n.pt format=onnx opset=13
```

And then deploy it using DeepSparse:

```python
from deepsparse import Pipeline

# Specify the path to your YOLOv8 ONNX model
model_path = "path/to/yolov8n.onnx"

# Set up the DeepSparse Pipeline
yolo_pipeline = Pipeline.create(task="yolov8", model_path=model_path)

# Run the model on your images
images = ["path/to/image.jpg"]
pipeline_outputs = yolo_pipeline(images=images)
```

Hope this helps! Let me know if there's anything else you need. 😊

**Comment by Framecyber** [comment]
I'm curious that can I use sparseML with the open ai clip model? is there a code or support for this?

**Reply by glenn-jocher** [reply]
@Framecyber SparseML primarily focuses on computer vision models including YOLO and ResNet, and I haven't seen specific examples for CLIP models in their documentation. While theoretically possible since SparseML works with PyTorch models, you'd likely need to create custom pruning recipes. I recommend checking Neural Magic's GitHub or reaching out to their team directly for the most current information on CLIP support.

**Comment by Nithinpbabu** [comment]
Can I use DeepSparse on Raspberry pi 5? (To process in CPU)

**Reply by glenn-jocher** [reply]
@Nithinpbabu Yes, DeepSparse can work on Raspberry Pi 5 since it's specifically designed for CPU optimization. However, performance will depend on your model size and the Pi's ARM architecture. DeepSparse is particularly effective on x86 CPUs with AVX-512 instructions, so while it will run on the Pi 5, you might not see the same dramatic speedups as on Intel/AMD systems. For the best Pi 5 performance, I'd recommend using a smaller YOLO11 model like YOLO11n exported to ONNX format.
